src/linker/Linker.Steps/SubStepsDispatcher.cs

```diff
@@ -0,0 +1,219 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+using Mono.Cecil;
+using Mono.Collections.Generic;
+
+namespace Mono.Linker.Steps
+{
+	//
+	// Generic steps dispatcher is intended to by used by custom linker step which
+	// consist of multiple steps. It simplifies their implementation as well as the
+	// way how to hook them into the pipeline of existing steps.
+	//
+	public abstract class SubStepsDispatcher : IStep, IEnumerable<ISubStep>
+	{
+		readonly List<ISubStep> substeps;
+
+		List<ISubStep> on_assemblies;
+		List<ISubStep> on_types;
+		List<ISubStep> on_fields;
+		List<ISubStep> on_methods;
+		List<ISubStep> on_properties;
+		List<ISubStep> on_events;
+
+		protected SubStepsDispatcher ()
+		{
+			substeps = new List<ISubStep> ();
+		}
+
+		protected SubStepsDispatcher (IEnumerable<ISubStep> subSteps)
+		{
+			substeps = new List<ISubStep> (subSteps);
+		}
+
+		public void Add (ISubStep substep)
+		{
+			substeps.Add (substep);
+		}
+
+		void IStep.Process (LinkContext context)
+		{
+			InitializeSubSteps (context);
+
+			BrowseAssemblies (context.GetAssemblies ());
+		}
+
+		static bool HasSubSteps (List<ISubStep> substeps) => substeps?.Count > 0;
+
+		void BrowseAssemblies (IEnumerable<AssemblyDefinition> assemblies)
+		{
+			foreach (var assembly in assemblies) {
+				CategorizeSubSteps (assembly);
+
+				if (HasSubSteps (on_assemblies))
+					DispatchAssembly (assembly);
+
+				if (!ShouldDispatchTypes ())
+					continue;
+
+				BrowseTypes (assembly.MainModule.Types);
+			}
+		}
+
+		bool ShouldDispatchTypes ()
+		{
+			return HasSubSteps (on_types)
+				|| HasSubSteps (on_fields)
+				|| HasSubSteps (on_methods)
+				|| HasSubSteps (on_properties)
+				|| HasSubSteps (on_events);
+		}
+
+		void BrowseTypes (Collection<TypeDefinition> types)
+		{
+			foreach (TypeDefinition type in types) {
+				DispatchType (type);
+
+				if (type.HasFields && HasSubSteps (on_fields))
+					BrowseFields (type.Fields);
+
+				if (type.HasMethods && HasSubSteps (on_methods))
+					BrowseMethods (type.Methods);
+
+				if (type.HasProperties && HasSubSteps (on_properties))
+					BrowseProperties (type.Properties);
+
+				if (type.HasEvents && HasSubSteps (on_events))
+					BrowseEvents (type.Events);
+
+				if (type.HasNestedTypes)
+					BrowseTypes (type.NestedTypes);
+			}
+		}
+
+		void BrowseFields (Collection<FieldDefinition> fields)
```

Comment:
Nit: I would personally inline these into the big Browse method - I know that JIT will probably inline them anyway, but still...

Reply:
ok